Add wavelength rangefinder

simonrp84 opened the pull request:
This PR adds a utility function that retrieves the minimum and maximum wavelengths in a given spectral response that are above a threshold value. It also returns the central wavelength:

```
seviri = RelativeSpectralResponse('Meteosat-8', 'seviri')
wvl_range = get_wave_range(seviri.rsr['HRV']['det-1'], 0.15)
print(wvl_range)
[0.456, 0.715, 0.996]
```

The main purpose for this function is to help with setting the wavelength limits in satpy reader YAML files, which are [currently somewhat arbitrary](https://github.com/pytroll/satpy/pull/1404).
I also added a utility script to the `./bin` directory: `find_wavelengths_for_satpy.py`. This script calls the above utility function and prints the wavelength range in a YAML-friendly formatting:
`python find_wavelengths_for_satpy.py meteosat-8 seviri `
Returns:
```
name:  VIS0.6
  wavelength: [0.599, 0.640, 0.683]
name:  IR8.7
  wavelength: [8.476, 8.711, 8.924]
name:  IR7.3
  wavelength: [7.050, 7.357, 7.670]
name:  IR13.4
  wavelength: [12.520, 13.351, 14.160]
name:  VIS0.8
  wavelength: [0.776, 0.809, 0.844]
name:  HRV
  wavelength: [0.456, 0.715, 0.996]
name:  IR6.2
  wavelength: [5.818, 6.306, 6.790]
name:  IR12.0
  wavelength: [11.400, 11.943, 12.520]
name:  IR10.8
  wavelength: [10.200, 10.788, 11.400]
name:  NIR1.6
  wavelength: [1.567, 1.635, 1.707]
name:  IR9.7
  wavelength: [9.526, 9.671, 9.817]
name:  IR3.9
  wavelength: [3.603, 3.920, 4.254]
```

 - [x] Tests added
 - [ ] Tests passed: Passes ``pytest pyspectral``
 - [x] Passes ``flake8 pyspectral``
 - [x] Fully documented

simonrp84 commented:
Just a note, this PR now also updated stickler-ci to function with Python 3 instead of Python 2

adybbroe reviewed (approved):
LGTM

pnuu commented on pyspectral/utils.py:
```diff
@@ -554,3 +551,24 @@ def get_logger(name):
     return log
+
+
+def get_wave_range(in_chan, threshold):
```
Should the `threshold` be a kwarg with a reasonable default value?

pnuu commented on bin/find_wavelengths_for_satpy.py:
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2020 Simon Proud
+
+# Author(s):
+
+#   simon.proud@physics.ox.ac.uk <simon.proud@physics.ox.ac.uk>
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
```
Please fill the empty lines with `#`. And I believe the current preferred copyright is `Pytroll developers`.

adybbroe replied:
Agree, can you fix @simonrp84 ?

pnuu commented on bin/find_wavelengths_for_satpy.py:
```diff
@@ -0,0 +1,66 @@
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description='Retrieve channel wavelength ranges for a given sensor.')
+    parser.add_argument("platform_name",
+                        help="The Platform name",
+                        type=str)
+    parser.add_argument("sensor",
+                        help="The sensor/instrument name",
+                        type=str)
+    parser.add_argument("-r", "--minimum_response",
+                        help=("Minimum response threshold: Defines the value "
+                              "that the RSR must exceed in order to trigger"
+                              "the minimum or maximum wavelengths."),
+                        default=0.15, type=float)
+    parser.add_argument("--bandname", '-b',
+                        help="The sensor band name. Leave blank to return all bands.", type=str)
+    parser.add_argument("--detector", '-d',
+                        help="The sensor detector, if not passed will default to det-1",
+                        default='det-1', type=str)
+    args = parser.parse_args()
```
Please put this in a separate function, `parse_args()` or something.

adybbroe replied:
yeah, right. I think @simonrp84 took from how I made it with the other scripts. But better yes to have a separate function.

pnuu commented on bin/find_wavelengths_for_satpy.py:
```diff
@@ -0,0 +1,66 @@
+                        default='det-1', type=str)
+    args = parser.parse_args()
+
+    platform = args.platform_name
+    sensor = args.sensor
+    threshold = args.minimum_response
+
+    rsr = RelativeSpectralResponse(platform, sensor)
+    if args.bandname:
+        bands = [args.bandname]
+    else:
+        bands = rsr.band_names
+    det = args.detector
+
+    for bname in bands:
+        wvls = get_wave_range(rsr.rsr[bname][det], args.minimum_response)
+        print(f'name:  {bname}')
+        print(f'  wavelength: [{wvls[0]:5.3f}, {wvls[1]:5.3f}, {wvls[2]:5.3f}]')
```
This should also be wrapped in a `main()` function that'd be called from the `if __name__ == "__main__"` block.

adybbroe replied:
:+1:

pnuu reviewed (approved):
Few suggestions, otherwise looks good to me.

simonrp84 commented:
Ok, I think I've fixed all of your suggestions / comments. I updated the copyright notice in (AFAIK) all the python files, so it should be consistent now, using `Pytroll Developers`.
The downside of updating the copyright is that codefactor complains about many files, and that test is failing. I don't intend to fix all the problems it raises, though.

So, unless there's any other issues that you see this should be ready now :)

simonrp84 commented:
Don't understand why Codacy is complaining, it wants a blank line in the docstring...and there is a blank line in the docstring?

djhoese commented:
Maybe it didn't rerun after you changed something? Looks right to me too.

djhoese commented:
I told codacy to reanalyze the code and it still had the same issue. Maybe add a blank line at the *end* of the docstring (between the last sentence and the `"""`) and see if that fixes it. I know it isn't what it is complaining about but :man_shrugging:

pnuu commented on pyspectral/utils.py:
```diff
@@ -554,3 +551,24 @@ def get_logger(name):
+    then drop back to near zero. This function takes advantage of this to find
+    the first and last points where the RSR is greater than a threshold. These
+    points are then defined as the minimum and maximum wavelengths for a
+    given channel, and can be used, for example, in Satpy reader YAML files.
```
```suggestion
    given channel, and can be used, for example, in Satpy reader YAML files.
    
```
This should fix the Codacy complaint.

adybbroe reviewed (approved):
:+1:

adybbroe commented:
Thanks a lot @simonrp84
